File: leetcode/2614/c/main.c
#include <math.h>
#include <stdbool.h>

#define max(a, b) (a) > (b) ? (a) : (b)
bool is_prime(int num) {
    if (num < 2) {
        return false;
    } else if (num == 2) {
        return true;
    }

    int upto = (int)(sqrt((double)num));

    for (int i = 2; i <= upto; i++) {
        if (num % i == 0) {
            return false;
        }
    }

    return true;
}

int diagonalPrime(int** nums, int numsSize, int* numsColSize) {
    int max = 0;

    for (int i = 0; i < numsSize; i++) {
        int idx1 = (numsSize + 1) * i;
        int idx2 = (numsSize * numsSize - 1) - (numsSize - 1) * (i + 1);
        int num1 = nums[idx1 / 3][idx1 % 3];
        int num2 = nums[idx2 / 3][idx2 % 3];

        if (is_prime(num1)) {
            max = max(num1, max);
        }

        if (is_prime(num2)) {
            max = max(num2, max);
        }
    }

    return max;
}

File: leetcode/2614/rs/src/main.rs
use std::collections::HashSet;

struct Solution;
impl Solution {
    pub fn diagonal_prime(nums: Vec<Vec<i32>>) -> i32 {
        let diag_idxs = Self::get_diagnal_idxs(nums.len() as i32);
        println!("{:?}", diag_idxs);
        nums.iter()
            .flatten()
            .enumerate()
            .filter_map(|(i, &num)| {
                if diag_idxs.contains(&(i as i32)) {
                    Some(num)
                } else {
                    None
                }
            })
            .filter(|&num| Self::is_prime(num))
            .max()
            .unwrap_or(0)
        // *nums.iter().flatten().filter(|&num| Self::is_prime(*num)).max().unwrap_or(&0)
    }

    fn get_diagnal_idxs(n: i32) -> HashSet<i32> {
        let mut idxs = HashSet::new();

        for i in 0..n {
            idxs.insert((n + 1) * i);
            idxs.insert((n * n - 1) - (n - 1) * (i + 1));
        }

        idxs
    }

    fn is_prime(num: i32) -> bool {
        if num < 2 {
            return false;
        } else if num == 2 {
            return true;
        }

        // `num` must not be divisible by any number from 2 upto sqrt(num).
        (2..=num.isqrt()).all(|k| num % k != 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_prime() {
        assert!(Solution::is_prime(2));
        assert!(Solution::is_prime(3));
        assert!(!Solution::is_prime(4));
        assert!(Solution::is_prime(5));
        assert!(!Solution::is_prime(6));
    }
}
fn main() {
    println!("Hello, world!");
}
